Source/CarthageKit/Project.swift

```diff
@@ -733,20 +742,52 @@ public final class Project {
 
 	/// Attempts to build each Carthage dependency that has been checked out,
 	/// optionally they are limited by the given list of dependency names.
+	/// Cached dependencies whose dependency trees are also cached will not
+	/// be rebuilt unless otherwise specified via build options.
 	///
 	/// Returns a producer-of-producers representing each scheme being built.
 	public func buildCheckedOutDependenciesWithOptions(_ options: BuildOptions, dependenciesToBuild: [String]? = nil, sdkFilter: @escaping SDKFilterCallback = { .success($0.0) }) -> SignalProducer<BuildSchemeProducer, CarthageError> {
 		return loadResolvedCartfile()
-			.flatMap(.merge) { resolvedCartfile in
+			.flatMap(.concat) { resolvedCartfile -> SignalProducer<Dependency<PinnedVersion>, CarthageError> in
 				return self.buildOrderForResolvedCartfile(resolvedCartfile, dependenciesToInclude: dependenciesToBuild)
 			}
+			.flatMap(.concat) { dependency -> SignalProducer<(Dependency<PinnedVersion>, Set<ProjectIdentifier>, Bool?), CarthageError> in
+				return SignalProducer.combineLatest(
+					SignalProducer(value: dependency),
+					self.dependencyProjects(for: dependency),
+					versionFileMatches(dependency, platforms: options.platforms, rootDirectoryURL: self.directoryURL)
+				)
+			}.reduce([]) { (includedDependencies, nextGroup) -> [Dependency<PinnedVersion>] in
```

Comment:
I tried replacing `nextGroup` with a tuple as you suggest but I couldn't get it to compile.